Automatically show closed captioning/hearing impaired text track

xirac opened the pull request:
Following #6055 and after reading #5749 , here are changes on `setPrerredTextLanguage` and `getFormatLanguageScore` allowing to automatically show closed captioning/hearing impaired text track by using the following trackSelectorParameter for instance:
`trackSelectorParameters=newDefaultTrackSelector.ParametersBuilder().setPreferredTextLanguage("fra",C.ROLE_FLAG_CAPTION).build();`

tonihei commented:
Thanks for taking a look at this!

I'll add some comments to the code for discussion.

tonihei commented on library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java:
```diff
@@ -420,6 +420,12 @@ public ParametersBuilder setPreferredTextLanguage(@Nullable String preferredText
     }
 
+    @Override
+    public ParametersBuilder setPreferredTextLanguage(@Nullable String preferredTextLanguage, int preferredRoleFlags) {
```
I think the API we had in mind when discussing this issue a while ago was having a separate method to set the preferred role flags.

This would allow the following combinations:
1. No language specified, no role flags => select no text track unless marked as "default" or "forced".
3. No language specified, but role flags specified => select text track with matching or partly matching role flags and the selected audio language, select none (or the default/forced track) if no text track with these flags and language is available
2. Language specified, no role flags => select text track of the best matching language, prefer default, non-forced and  tracks without caption flags. If no track with this language is available, select "default" or "forced" in other language if possible.
4. Language and role flags specified => select text track of the best matching language, prefer track with matching or partly matching role flags. If not available, fall back to default, non-forced tracks of the preferred language if possible. If no track of the language is available, fall back to matching role flags in other languages, or default or forced tracks in other languages.

So if someone wants to integrate with CaptioningManager, they would use it like that:
```
if (CaptioningManager.isEnabled()) {
  paramsBuilder.setPreferredRoleFlags(C.ROLE_FLAG_CAPTION | C.ROLE_FLAG_DESCRIBES_MUSIC_AND_SOUND);
  Locale preferredLocale = CaptioningManager.getLocale();
  if (preferredLocale != null) {
    paramsBuilder.setPreferredTextLanguage(Util.getLocaleLanguageTag(preferredLocale));
  }
}
```
This method is complicated enough that we should probably integrate it into SimpleExoPlayer for convenience (or document it somewhere). Not in this PR, that's a just a general reminder for us to do this :)

tonihei commented on library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java:
```diff
@@ -2079,7 +2089,7 @@ private static boolean isSupportedAdaptiveAudioTrack(
           boolean isForced = (maskedSelectionFlags & C.SELECTION_FLAG_FORCED) != 0;
           int trackScore;
-          int languageScore = getFormatLanguageScore(format, params.preferredTextLanguage);
+          int languageScore = getFormatLanguageScore(format, params.preferredTextLanguage, params.preferredRoleFlags);
```
The track score logic in selectTextTrack gets increasingly complicated, especially if we follow the proposal I wrote down above. I think before continuing with this change, it would be best to switch from the score based logic to a comparison based logic similar to the one we use for audio track selection (see AudioTrackScore). This allows to get rid of these seemingly arbitrary score numbers (they are not arbitrary, but difficult to understand) and also make the logic more straight-forward. This first change should be a separate PR without changing the actual result and the new functionality here can be then added on top. Are you planning to take a look at this? Otherwise you can wait until we get around to do that ourselves and we'll revisit this PR once that is done.

xirac replied:
Thanks for your comment.
Here is the separate PR moving logic to a TextTrackScore #6178 .

ojw28 commented:
Is the plan to update this now that #6178 has been merged? Thanks!

xirac commented:
Thanks for the merge of #6178 . I plan to update this one within next 2 weeks.

Commit:
Automatically show closed captioning/hearing impaired text track

xirac commented:
Here is the PR update using TextTrackScore.

Commit:
Missing documentation link

Commit:
Merge remote-tracking branch 'upstream/dev-v2' into dev-v2

# Conflicts:
#	library/core/src/main/java/com/google/android/exoplayer2/trackselection/TrackSelectionParameters.java

xirac commented:
Do you need some code changes on this pull request?

ojw28 commented:
@tonihei - PTAL

tonihei commented on library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java:
```diff
@@ -2702,6 +2713,7 @@ public TextTrackScore(
       preferredLanguageScore =
           getFormatLanguageScore(
               format, parameters.preferredTextLanguage, parameters.selectUndeterminedTextLanguage);
+      preferredRoleFlagsScore = format.roleFlags & parameters.preferredRoleFlags;
```
I think you actually need to count the number of matches: Integer.bitCount(format.roleFlags & parameters.preferredRoleFlags)

tonihei commented on library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java:
```diff
@@ -2712,7 +2724,7 @@ public TextTrackScore(
           getFormatLanguageScore(format, selectedAudioLanguage, selectedAudioLanguageUndetermined);
       isWithinConstraints =
-          preferredLanguageScore > 0 || isDefault || (isForced && selectedAudioLanguageScore > 0);
+          (preferredLanguageScore > 0 || (parameters.preferredTextLanguage == null && selectedAudioLanguageScore > 0 && preferredRoleFlagsScore > 0)) || isDefault || (isForced && selectedAudioLanguageScore > 0);
```
I'm not sure the `selectedAudioLanguageScore > 0` part makes sense. If someone specifies preferred role flags without providing a preferred language, we should probably still select a track with the matching role flags if available.

tonihei commented on library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java:
```diff
@@ -2730,6 +2742,9 @@ public int compareTo(TextTrackScore other) {
       if (this.preferredLanguageScore != other.preferredLanguageScore) {
         return compareInts(this.preferredLanguageScore, other.preferredLanguageScore);
       }
+      if (this.preferredRoleFlagsScore != other.preferredRoleFlagsScore) {
```
If everything else is exactly the same, and no role flag preference is provided in the parameters, we should probably still compare by the existing role flags and prefer the tracks *without* CAPTION or DESCRIBES_MUSIC_AND_SOUND role flags. This under the assumption that no specification means that non-captioning text tracks are preferred.  This comparison should be the very last one after the selectedAudioLanguageScore.

tonihei commented:
Sorry for the delay. The PR looks good in general, I just left some suggestions for improvement.

Commit:
Count the number of matches for role flags.
Select a track with matching role flags even if someone specifies a preferred role flgs without providing a preferred language
Prefer track without captions or describing music and sound flag when no role flag preference is provided.

xirac commented:
Thanks. I pushed the 3 changes corresponding to your suggestions.

tonihei commented on library/core/src/main/java/com/google/android/exoplayer2/trackselection/DefaultTrackSelector.java:
```diff
@@ -2736,7 +2753,17 @@ public int compareTo(TextTrackScore other) {
+      if (!this.hasPreferredRoleFlags) {
+        return compareInts(other.hasCaptionRoleFlags, this.hasCaptionRoleFlags);
+      }
+      return compareInts(this.preferredRoleFlagsScore, other.preferredRoleFlagsScore);
```
I actually meant that this comparison stays where it was (after the language score), but never mind, I just move it up again when merging.

tonihei commented:
Change is merged now. Thanks again for all the work!

xirac commented:
Thanks for the merge, you're welcome.
